lib/Transforms/IPO/StructFieldAccessInfo.cpp

```diff
@@ -0,0 +1,343 @@
+  // structs
+  assert(Op->getType()->isIntegerTy());
+  // Operand 2 should be the index to the field, and be a constant
+  Op = U->getOperand(2);
+  assert(isa<Constant>(Op));
+  auto *Index = dyn_cast<Constant>(Op);
+  auto Offset = (FieldNumType)Index->getUniqueInteger().getZExtValue();
+  assert(Offset < NumElements);
+  // TODO: ignore indices after this one. If there's indices, the field has to
+  // be an array or struct
+  return Offset + 1; // return field number starting from 1
+}
+
+void StructFieldAccessInfo::addFieldAccessFromGEP(const User *U) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_IR,
+                  dbgs() << "Analyze all users of GEP: " << *U << "\n");
+  assert(isa<GetElementPtrInst>(U) || isa<GEPOperator>(U));
+  auto FieldLoc = calculateFieldNumFromGEP(U);
+  if (FieldLoc == 0)
+    return;
+  for (auto *User : U->users()) {
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR,
+                    dbgs() << "Check user of " << *U << ": " << *User << "\n");
+    if (isa<Instruction>(User)) {
+      auto *Inst = dyn_cast<Instruction>(User);
+      if (Inst->getOpcode() == Instruction::Load)
+        addFieldAccessNum(Inst, FieldLoc);
+      else if (Inst->getOpcode() == Instruction::Store) {
+        if (U == Inst->getOperand(1))
+          addFieldAccessNum(Inst, FieldLoc);
+      } else {
+        if (Inst->getOpcode() == Instruction::Call) {
+          auto *Call = dyn_cast<CallInst>(Inst);
+          auto *Func = Call->getCalledFunction();
+          if (Func && Func->arg_size() == Call->getNumArgOperands()) {
+            for (unsigned i = 0; i < Call->getNumArgOperands(); i++) {
+              if (Call->getArgOperand(i) == U)
+                addFieldAccessNum(Inst, Func, i, FieldLoc);
+            }
+          } else {
+            // Bail out if Gep is used in an indirect function or a function
+            // that has undetermined args
+            addStats(StructFieldAccessManager::DebugStats::
+                         DS_GepPassedIntoIndirectFunc);
+          }
+        } else if (Inst->getOpcode() == Instruction::Invoke) {
+          auto *Call = dyn_cast<InvokeInst>(Inst);
+          auto *Func = Call->getCalledFunction();
+          if (Func) {
+            for (unsigned i = 0; i < Call->getNumArgOperands(); i++) {
+              if (Call->getArgOperand(i) == U)
+                addFieldAccessNum(Inst, Func, i, FieldLoc);
+            }
+          } else {
+            addStats(StructFieldAccessManager::DebugStats::
+                         DS_GepPassedIntoIndirectFunc);
+          }
+        } else if (Inst->getOpcode() == Instruction::BitCast) {
+          addStats(
+              StructFieldAccessManager::DebugStats::DS_GepPassedIntoBitcast);
+        } else {
+          // TODO: Collect stats of this kind of access and add analysis later
+          addStats(StructFieldAccessManager::DebugStats::DS_GepUnknownUse,
+                   Inst->getOpcode());
+        }
+      }
+    } else if (isa<Operator>(User)) {
```

Comment:
Please let me know if I understand this correctly. Do you mean instead of just checking if User is a Instruction or Operator, I also check if it can be a ConstantExpr and add stats for it?